vault/identity_store_oidc_provider.go

```diff
@@ -2545,51 +2551,83 @@ func (i *IdentityStore) storeOIDCDefaultResources(ctx context.Context, view logi
 		i.Logger().Debug("wrote OIDC default provider")
 	}
 
-	// Store the default key
-	storageKey = namedKeyConfigPath + defaultKeyName
+	if _, err := i.ensureDefaultKey(ctx, view); err != nil {
+		return fmt.Errorf("error writing default key to storage: %w", err)
+	}
+
+	// Store the allow all assignment
+	storageKey = assignmentPath + allowAllAssignmentName
 	entry, err = view.Get(ctx, storageKey)
 	if err != nil {
 		return err
 	}
 	if entry == nil {
-		defaultKey := defaultOIDCKey()
-
-		// Generate initial key material for current and next keys
-		err = defaultKey.generateAndSetKey(ctx, i.Logger(), view)
-		if err != nil {
-			return err
-		}
-		err = defaultKey.generateAndSetNextKey(ctx, i.Logger(), view)
-		if err != nil {
-			return err
-		}
-
-		// Store the entry
-		entry, err := logical.StorageEntryJSON(storageKey, defaultKey)
+		entry, err := logical.StorageEntryJSON(storageKey, allowAllAssignment())
 		if err != nil {
 			return err
 		}
 		if err := view.Put(ctx, entry); err != nil {
 			return err
 		}
-		i.Logger().Debug("wrote OIDC default key")
+		i.Logger().Debug("wrote OIDC allow_all assignment")
 	}
 
-	// Store the allow all assignment
-	storageKey = assignmentPath + allowAllAssignmentName
-	entry, err = view.Get(ctx, storageKey)
+	return nil
+}
+
+func (i *IdentityStore) ensureDefaultKey(ctx context.Context, storage logical.Storage) (*namedKey, error) {
```

Comment:
Maybe worth a comment to call out that this function doesn't actually generate the key, only its config. I would suggest a different name to convey that instead, but it's quite a subtle distinction to make.

Reply:
Good call! Added comments in https://github.com/hashicorp/vault/pull/24332/commits/3c3af4b92e497b4c5a70a70af3cdcdb6881dc38e.